File: java/starter/acquirer/src/main/java/network/t0/pay/acquirer/internal/Times.java
package network.t0.pay.acquirer.internal;

import com.google.protobuf.Timestamp;

import java.time.Instant;

/**
 * Timestamps on the wire are absolute UTC.
 *
 * <p>Use {@link #format} in logs: protobuf's own {@code Timestamp.toString()} is
 * multi-line text format and turns one log line into six.
 */
public final class Times {

    public static Timestamp from(Instant instant) {
        return Timestamp.newBuilder()
                .setSeconds(instant.getEpochSecond())
                .setNanos(instant.getNano())
                .build();
    }

    public static Instant toInstant(Timestamp timestamp) {
        return Instant.ofEpochSecond(timestamp.getSeconds(), timestamp.getNanos());
    }

    public static String format(Timestamp timestamp) {
        return toInstant(timestamp).toString();
    }

    private Times() {
    }
}

File: go/starter/acquirer/internal/times.go
package internal

import (
	"time"

	"google.golang.org/protobuf/types/known/timestamppb"
)

// FormatTimestamp formats a protobuf Timestamp as an ISO 8601 string.
func FormatTimestamp(ts *timestamppb.Timestamp) string {
	if ts == nil {
		return "<nil>"
	}
	return ts.AsTime().UTC().Format(time.RFC3339Nano)
}
